### PR TITLE
Update FSE template part hover styles

### LINKED ISSUE #34659: FSE: Update the design of the header hover
The design of the header on hover needs more polish. It's too dark to see through on hover, and too quick to highlight an unhighlight making it disorienting.

TL;DR from conversation in https://github.com/Automattic/wp-calypso/pull/35158
====== 
Let's update the hover state to:
62586343-05869780-b872-11e9-80f5-639a6a171733

And the Focus State to the following possibly using https://caniuse.com/#feat=css-backdrop-filter
62586353-146d4a00-b872-11e9-9aac-20e4c31dbf61

### PR DESCRIPTION
**Pull request by jeraldjuice**
### Summary

This PR changes some of the functionality and design of the template part hover actions to make it less jarring. 

Updated-FSE-header-hover

<hr>

### Specifics

* Old hover styles removed. New hover styles are a gray left border and allowing the breadcrumb to be visible.
* Removed the functionality to disable selecting template blocks. Instead, duplicated it inside of the template block component to ensure that the sidebar is closed if the block is selected/unselect the block when the sidebar is opened.
* Placeholder is only visible when the block is selected.

<hr>

### Screenshots

**Hover state**

Screen Shot 2019-08-07 at 1 34 47 PM

**Focus state**

Screen Shot 2019-08-07 at 1 34 54 PM

<hr>

### Testing

* Setup FSE using your local development environment of choice.
* Open up the new page editor. 
* Ensure that the hover styles are visible when hovering a template block, but not for the post content slot.
* Click on a template part block. Ensure that the placeholder is visible.
* Ensure that it is not possible to have the placeholder visible and the settings sidebar open at the same time.

Fixes #34659

### CODE REVIEW

**Comment by shaunandrews** [comment]
I think the PR needs a rebase.

--

For the hover lets keep it as "stock Gute" as possible. It should look and act just like any other Block on the page. Here's what I'd expect to see when hovering the header:

image

If we can customize the label there, it'd be great to use the template_part name (in this example "Header") somewhere, rather than just the generic "Template" label.

--

Instead of using the dialog-like UI with the "...appear on all of your site's pages." message, lets use the standard Block toolbar and add our own button.

image

I've made the contents of the block translucent when focused to help make it clear its not editable.

--

I noticed if you have the Sidebar open and then click on a template_part, the Sidebar inexplicably closes. It should stay open, and we should show a Block description and the button to edit.

image

**Comment by apeatling** [comment]
@shaunandrews Give master a try, this UI is no longer present for template and content blocks, which presents a challenge on the above.

**Comment by jeraldjuice** [comment]
@shaunandrews Yes, as @apeatling mentioned, we've gotten rid of a lot of the UI around these blocks recently, which presents a bit of an obstacle towards making it look similarly to traditional blocks.

The other thing that's in play is some extensibility issues with Gutenberg - primarily, restricting editing template part block settings. Previously, this was implemented by not allowing the user to select the template part block (which meant that accessing the settings wasn't possible). I've removed that and surfaced the functionality a bit, now allowing a user to select the template part block in order to show the message, but not allowing the settings sidebar to be opened at the same time. (Ie, if the sidebar is opened and you select a template part block, the sidebar closes. If a template part block is selected and you open the sidebar, the block deselects.)

**Comment by noahtallen** [comment]
One note with some of the work I did (which seems to have been premature :p) is that we would have to find a new way to handle `customClassName` support. I don't think we want it to show up as a setting to change, but for the moment it has to in order to support how @Copons handled putting a custom className around the header/footer. I think we could probably put that into our new approach where we add the className with the gutenberg hook:
https://github.com/Automattic/wp-calypso/blob/b20b7ea0812694f4041352a02a9a7fe9848a180a/apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/index.js#L37-L47

Then we can turn off the setting and (if we want the block UI to show up) enable the selection again without seeing unwanted settings

**Comment by noahtallen** [comment]
Just an FYI so we don't do the same thing at once, I'm planning on reworking the className a bit as part of https://github.com/Automattic/wp-calypso/issues/35059, and I'll make sure to turn of the className setting so that you can do whatever's needed for this without exposing any unwanted settings :)

**Comment by shaunandrews** [comment]
> this UI is no longer present for template and content blocks, which presents a challenge on the above.

Sorry if I'm being confusing. I was suggesting we bring it back in a limited fashion (or add or own variation), but make it actually useful. Without the button, the UI is useless. But, if we can add the button, this UI feels like the most natural place for it.

**Comment by apeatling** [comment]
Had a back and forth with Shaun about this. My perspective is that using the block UI for this experience is going to result in folks missing the button. As a user I would expect to be able to click the site title block, and my eye is focused on that, not on the edit header button in the top left.

Shaun's perspective is that it follows the already existing Gutenberg patterns. He's right about that, I just think we need to get in the way of users more for this because it's not going to be doing what they expect (direct manipulation).

My expectation is something more like this that uses a similar interaction model but it just less intense (we don't need the message):

Screen Shot 2019-08-06 at 3 07 55 PM

https://developer.mozilla.org/en-US/docs/Web/CSS/backdrop-filter

**Comment by shaunandrews** [comment]
To be clear, the mockup Andy posted above would be for the focus state. The :hover state should be the same as any other block. That said, I'd be fine with it — but if we want to run it by people in the usability tests, we wouldn't know if the default styling (with the button in the toolbar) would be suitable or not.

**Comment by gwwar** [comment]
> To be clear, the mockup Andy posted above would be for the focus state. The :hover state should be the same as any other block.

@shaunandrews If we have the following as focus state:
62550184-cf2c2680-b837-11e9-9513-cc9efa6f45c1

Won't this mean that we'll need two clicks to get to the header editor? In:

62580821-17f5d680-b85c-11e9-87b4-1c6e28653c06

**Comment by shaunandrews** [comment]
> we'll need two clicks to get to the header editor?

Yup, that's correct. The alternative is to show the button when :hovered, and that just feels too obtrusive to me. I could be convinced otherwise, but this is the standard action for selecting blocks in Gutenberg.

**Comment by gwwar** [comment]
Thanks @shaunandrews for the clarification. I updated https://github.com/Automattic/wp-calypso/issues/34659 to reflect the new requirements

**Comment by gwwar** [comment]
@apeatling looking at https://caniuse.com/#feat=css-backdrop-filter We usually target the last two major versions of popular browsers. There doesn't appear to be good support for this, but I'm sure we can find alternative for a blur effect. There should be a few other techniques.

**Comment by shaunandrews** [comment]
> find alternative for a blur effect

Don't worry about the blur — any use of CSS for blurring is going to cause browser issues. For now, just a straight-up overlay is good enough.

**Comment by jeraldjuice** [comment]
Thanks for all of the next steps on this! I'll change those states to reflect the new mockups.

**Comment by apeatling** [comment]
I think the blur could be applied and if the browser doesn’t support it fall back to the overlay by default. The background blur is supported natively in chrome and WebKit and I’ve found performance to be very good in other projects.

**Comment by Copons** [comment]
@jeraldjuice can this be of any interest for this PR?

https://github.com/Automattic/wp-calypso/pull/35190#issuecomment-519162908

> Are we aware that the image block's placeholder buttons have a z-index above the template part overlay?
>
> Aug-07-2019 17-02-34

**Comment by jeraldjuice** [comment]
@Copons I actually noticed that when testing Glen's PR, so I'm going to slip in a z-index. Thanks!

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/edit.js [hunk 1]
```diff
@@ -93,6 +102,20 @@ const TemplateEdit = compose(
 			receiveTemplateBlocks();
 		} );
 
+		useEffect( () => {
```
Comment (gwwar) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/edit.js]: For reviewers: https://reactjs.org/docs/hooks-effect.html 👍

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/edit.js [hunk 2]
```diff
@@ -65,6 +66,8 @@ const TemplateEdit = compose(
 				receiveBlocks( [ templateBlock ] );
 				setState( { templateClientId: templateBlock.clientId } );
 			},
+			closeGeneralSidebar: dispatch( 'core/edit-post' ).closeGeneralSidebar,
+			clearSelectedBlock: dispatch( 'core/editor' ).clearSelectedBlock,
```
Comment (gwwar) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/edit.js]: Optional, minor: noting that most of the code I've seen so far tends to pull this out into a `const`. I don't have strong feelings on this, but it's nice to see consistent code in a project.

eg:
```jsx
const { closeGeneralSidebar } = dispatch( 'core/edit-post' );
const { clearSelectedBlock } = dispatch( 'core/editor' );
//...
closeGeneralSidebar,
clearSelectedBlock
```

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/style.scss [hunk 3]
```diff
@@ -52,26 +55,15 @@
...
-					}
-				}
+			& > .editor-block-list__block-edit.block-editor-block-list__block-edit::before {
+				background-color: rgba( #000000, 0.5 );
+				z-index: 1;
 			}
 		}
+
+		&:not( .is-selected ):hover {
+			outline: #e2e4e7 solid;
```
Comment (gwwar) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/style.scss]: For follow up outside of this PR I wonder if it'd be good to start a named colors file.

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/style.scss [hunk 4]
```diff
@@ -52,26 +55,15 @@
...
-						outline: none;
-						box-shadow: none;
-					}
-					> .block-editor-block-list__breadcrumb {
-						display: none;
-					}
-				}
+			& > .editor-block-list__block-edit.block-editor-block-list__block-edit::before {
+				background-color: rgba( #000000, 0.5 );
+				z-index: 1;
```
Comment (gwwar) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/style.scss]: Thanks for the low z-index number 🙇‍♀ 

For folks 👀 I did make this extension if y'all had more complex cases to debug.
https://github.com/gwwar/z-context

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/edit.js (hunk 2 above)
Comment (jeraldjuice) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/edit.js]: Will do!

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/style.scss (hunk 3 above)
Comment (jeraldjuice) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/template/style.scss]: I definitely agree, especially as a way of consolidating a lot of the styling hacks we're doing to gutenify this.

**Comment by jeraldjuice** [comment]
Alright! I've consolidated the feedback I've gotten on this into some styling changes.

New hover state:

Screen Shot 2019-08-07 at 1 34 47 PM

New focus state:

Screen Shot 2019-08-07 at 12 12 20 PM

Focus state when blur isn't supported:

Screen Shot 2019-08-07 at 12 11 29 PM

I've kept the restricted no-sidebar-when-template-block-selected functionality until such a point as we fix the custom class options for the block.

**Comment by vindl** [comment]
A little bit late to the party, but nice work here @jeraldjuice! 🎉

**Comment by shaunandrews** [comment]
Very nice. Thanks for the changes!

**Pull request merged**